File: motion_planning/scripts/RRT.py
import math
import random
import copy
import matplotlib.pyplot as plt

show_animation = True

class RRT:

    def __init__(self, start, goal, obstacle_list, max_length, random_area):
        # start: starting location
        # goal: goal location
        # obstacle_list: list of obstacles
        # max_length: length of the nodes

        self.start = Node(start[0], start[1])
        self.goal = Node(goal[0], goal[0])
        self.obstacle_list = obstacle_list
        self.max_length = max_length        
        self.min_rand = random_area[0]
        self.max_rand = random_area[1]
    
    def path_planning(self, animation = True):
        # setting the list of nodes to the starting location
        self.nodes = [self.start]

        while True:

            # Random Sampling within the area of the grid
            # 80% of a nearby area on anywhere on the grid
            if random.randint(0, 100) > 20:
                rnd = [random.uniform(self.min_rand, self.max_rand), random.uniform(
                    self.min_rand, self.max_rand)]
            else:
                rnd = [self.goal.x, self.goal.y]

            # finding node index in relation to the goal
            new_ind = self.getNearestIndex(self.nodes, rnd) 
    
            # place the index into a new node to generate the new node
            nearestNode = self.nodes[new_ind]
            theta = math.atan2(rnd[1] - nearestNode.y, rnd[0] - nearestNode.x)

            newNode = copy.deepcopy(nearestNode)
            newNode.x += self.max_length * math.cos(theta)
            newNode.y += self.max_length * math.sin(theta)
            newNode.parent = new_ind

            # check if the node is in an obstacle
            # if not, add to list of nodes
            if not self.checkCollisions(newNode, self.obstacle_list):
                continue
            self.nodes.append(newNode)
            print("Number of Nodes so far: ", len(self.nodes))

            if animation:
                self.drawGraph()

            # check if the node hit the goal
            if self.checkGoal(newNode, self.goal):
                self.nodes.append(self.goal)
                if animation:
                    self.drawGraph()
                print("We hit the goal!")
                break
        
        # creating the final path that hit the goal

        # starting from the goal node and working backwards
        path = [[self.goal.x, self.goal.y]]
        last_ind = len(self.nodes) - 1

        # loops through each parent and appends to list of path
        while self.nodes[last_ind].parent is not None:
            node = self.nodes[last_ind]
            path.append([node.x, node.y])
            last_ind = node.parent
        
        # including the starting location
        path.append([self.start.x, self.start.y])

        return path


    def checkCollisions(self, node, obstacle_list):
        # not completely sure how the obstacle_list will work but I'm imaginging
        # something like a square with position and sizing
        for (x, y, size) in obstacle_list:
            dx = x - node.x
            dy = y - node.y
            d = math.sqrt(dx**2 + dy**2)
            if d <= (size):
                return False
            return True
    
    def checkGoal(self, node, goal):
        dx = node.x - goal.x
        dy = node.y - goal.y
        d = math.sqrt(dx**2 + dy**2)
        if d <= self.max_length:
            return True

    def getNearestIndex(self, nodes, rnd):
        distance = [(math.sqrt((node.x - rnd[0]) ** 2 + (node.y - rnd[1])** 2)) for node in nodes]
        min_index = distance.index(min(distance))
        return min_index

    def drawGraph(self):
        for node in self.nodes:
            if node.parent is not None:
                plt.plot([node.x, self.nodes[node.parent].x], [node.y, self.nodes[node.parent].y], "-g")
        
        for (x, y, size) in self.obstacle_list:
            plt.plot(x, y, "sk", ms = size*20)
        
        plt.plot(self.start.x, self.start.y, "or")
        plt.plot(self.goal.x, self.goal.y, "or")
        plt.axis([0, 20, 0, 20])
        plt.grid(True)
        plt.pause(0.01)

class Node:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.parent = None


def main():
    # set the start, goal, obstacles
    # everything needed to graph (still working on it)
    print("RRT")

    obstacleList = [(5,5,2)]

    rrt = RRT(start = [0, 0], goal = [10, 10], obstacle_list = obstacleList, max_length = 1, random_area=[0, 20])
    grid = rrt.path_planning(animation=show_animation)

    if show_animation:
        rrt.drawGraph()
        plt.plot([x for (x, y) in grid], [y for (x, y) in grid], '-r')
        plt.grid(True)
        plt.show()

if __name__ == '__main__':
    main()

File: motion_planning/scripts/RRT.m
%% Points

nrows = 500;
ncols = 500;

obstacle = false(nrows,ncols);

[x,y] = meshgrid (1:ncols, 1:nrows);

%% Obstacles

%creating obstacles
obstacle (y<10 | y>490 | x<10 | x>490) = true; % boundaries
obstacle (y > 150 & y < 250 & x > 150 & x < 250) = true;

figure;
imshow(~obstacle);

axis ([0 ncols 0 nrows]);
axis xy; %flips the y axis over

axis on;
xlabel ('x axis');
ylabel ('y axis');

title('Graph');
grid on

%% RRT Parameters

Max_Connect_Length = 500; % length of the node
Goal_found = false; % setting to false at first
Segments = 100; % Using segmentation to check for obstacles
                % Divide the edge between new node and neighbors into segments
                % Checking if all points on the line are away from
                % STILL WORKING ON THIS NOT COMPLETE YET
                % obstacles 

%% RRT Algorithm

Node_count = 1; % setting node counter to 1 at first for starting node
map = zeros(size(obstacle)); % setting the whole map to 0
map(obstacle) = 1; % setting the obstacles as 1
Tree_Connections = []; % holding the nodes and the tree

% Start Node
Start_node = [450, 450]; % We're assuming a starting node
Tree_Connections = [sub2ind(size(map), Start_node(1), Start_node(2)) Inf];
    % Adding the start node to the tree array
map(Start_node(2), Start_node(1)) = 2; % Marking the node as a 2
hold on
plot(Start_node(2), Start_node(1), 'b^')
hold off

% Goal Node
Goal_node = [50, 50];
hold on
plot(Goal_node(2), Goal_node(1), 'mv')
hold off

while(true)
        %% generating node at a random location
        Node_X = randi(ncols);
        Node_Y = randi(nrows);
        
        % Check if is through obstacle
        if (map(Node_Y, Node_X) == 1 || map(Node_Y, Node_X) == 2)
            continue;
        end
        
        %% Connect new node to closest node
        % For loop checks nodes within the max_length and puts in
        % neighbor_distances
        neighbor_distances = [];
        for i = 1: numel(Tree_Connections(:,1)) % 1 to number of elements in Tree_Connections[nodes]
            if (Tree_Connections(i,1) == Inf)
                break
            end
            [row, col] = ind2sub(size(map),Tree_Connections(i,1));
            
            % checks if in range
            if (norm([Node_Y, Node_X]-[row,col]) > Max_Connect_Length)
                continue
            end
            
            % ADD SOMETHING WITH SEGMENTATION
            
            % adding the node as a connection node (with the ID and the
            % distance to the new node)
            neighbor_distances = [neighbor_distances; [Tree_Connections(i,1), norm([Node_Y, Node_X]-[row,col])]];
                
        end
        
        %% Choose the closest node to connect to
        if (size(neighbor_distances) > 0)
            
            % Add node to the map and set parent to closest node
            map(Node_Y, Node_X) = 2;
            Node_count = Node_count + 1;
            distances_sorted = sortrows(neighbor_distances, 2);
            Tree_Connections = [Tree_Connections; sub2ind(size(map), Node_Y, Node_X) distances_sorted(1,1)];
            % adding the node from the distances_sorted because choosing
            % from the last part of the sorted array
            
            % Plotting the new node
            hold on
            plot(Node_X, Node_Y, 'r*')
            hold off
            
            % Plot edge between new node and its parent
            [row, col] = ind2sub( size(map), distances_sorted(1,1));
            hold on
            line([Node_X, col], [Node_Y,row])
            hold off
            
        else
            continue;
        end
        
        %% checking if we are close to the goal
        placeholder = [Node_Y, Node_X]-Goal_node;
        if(norm(placeholder) < Max_Connect_Length)
            Node_count = Node_count + 1;
            
            % Adding the goal_node to map
            Tree_Connections(Node_count, 1) = sub2ind(size(map), Goal_node(1), Goal_node(2));
            % Setting goal node parent to the new node
            Tree_Connections(Node_count,2) = sub2ind(size(map), Node_Y, Node_X);
            map(Goal_node(1),Goal_node(2))=2;
            
            %plot edge between node and goal
            hold on
            line([Node_X, Goal_node(2)], [Node_Y, Goal_node(1)])
            hold off
            
            Goal_found = true;
            
        end
        
        %% mark the path
        if (Goal_found)
            route = [];
            route = make_path(map, Nodes, Tree_Connections)
            % STILL WORKING ON MAKING THE PATH
            % plot the route
        end
        
        %% break if goal found
        if (Goal_found)
            break
        end

end
